pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -886,6 +888,35 @@ public void redeliverUnacknowledgedMessages() {
     }
 
     @Override
+    public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
+        if (!conf.isPerMessageRedeliverUnacknowledged()
+                || messageIds.size() > conf.getMaxRedeliverUnacknowledgedMessagesBatch()) {
+            redeliverUnacknowledgedMessages();
+            return;
+        }
+        ClientCnx cnx = cnx();
+        if (isConnected() && cnx.getRemoteEndpointProtocolVersion() >= ProtocolVersion.v2.getNumber()) {
+            MessageIdData.Builder builder = MessageIdData.newBuilder();
+            List<MessageIdData> messageIdDatas = messageIds.stream()
+                    .map(messageId -> {
+                        builder.setPartition(messageId.getPartitionIndex());
+                        builder.setLedgerId(messageId.getLedgerId());
+                        builder.setEntryId(messageId.getEntryId());
```

Comment:
@rdhabalia @merlimat I'm not really sure how we should deal with `BatchMessageIdImpl`. I believe it doesn't really make sense te redeliver a single message in a batch, I don't really know if it's possible too. So, what should we do when `UnAckedMessageTracker` considers a `BatchMessageIdImpl` should be redelivered?